### ISSUE TITLE
Issue: Modernize config.py for Full Hydra-based Configuration Management

### ISSUE THREAD

**Issue by huguesva** [rbalestr-lab/stable-pretraining#268]
The current config.py provides basic configuration dataclasses but doesn't fully leverage Hydra's capabilities
for instantiating and composing complex ML pipelines. The library needs a more robust configuration system that
allows users to:

1. Launch complete training jobs directly from Hydra config files
2. Instantiate all components (models, datasets, transforms, optimizers, schedulers, etc.) from configuration
3. Compose and override configurations easily for different experiments
4. Support config groups for modular configuration management

Current Limitations

- instantiate_config() only handles trainer instantiation, not other components
- Config dataclasses don't support recursive instantiation patterns
- No clear patterns for defining instantiable components in configs
- Missing support for common Hydra patterns (config groups, defaults, interpolation)
- Limited documentation on how to structure configs for complete experiments

Proposed Changes

1. Extend Config Classes to support full component instantiation:
@dataclass
class DataConfig:
    _target_: str  # e.g., "stable_pretraining.data.datasets.CIFARDataset"
    data_dir: str
    transform: dict  # nested config for transforms
    batch_size: int
    num_workers: int

@dataclass 
class ModelConfig:
    _target_: str  # e.g., "stable_pretraining.models.resnet.ResNet18"
    num_classes: int
    pretrained: bool = False

2. Update instantiate_config() to recursively instantiate all components:
  - Models, datasets, transforms, optimizers, schedulers
  - Support for partial instantiation and factory patterns
  - Handle dependencies between components
 
3. Support Hydra Best Practices:
  - Use _target_ for instantiation
  - Support _partial_ for factory functions
  - Enable config interpolation (${...})
  - Structured configs with validation

Example Usage After Update

# config.yaml
defaults:
  - model: resnet18
  - dataset: cifar10
  - optimizer: sgd
  - scheduler: cosine

trainer:
  _target_: stable_pretraining.trainer.Trainer
  epochs: 100
  device: cuda

# Command line usage:
# python train.py model=vit_small dataset=imagenet optimizer.lr=0.001

Benefits

- Reproducibility: Complete experiment configuration in YAML files
- Flexibility: Easy experimentation through config overrides
- Modularity: Swap components without code changes
- Type Safety: Structured configs with validation
- Community Standard: Follows Hydra/OmegaConf best practices

Implementation Tasks

- Refactor config dataclasses to support instantiation patterns
- Create comprehensive config templates for all major components
- Update instantiate_config() for recursive instantiation
- Add config validation and error handling
- Create example configs for common use cases
- Document configuration patterns and best practices
- Add tests for config instantiation flows

This update would make the library significantly more user-friendly and align with modern ML engineering
practices where configuration-driven development is the standard.

**Comment by huguesva** [comment]
solved in #270
